llvm/test/Transforms/FunctionSpecialization/ssa-copy.ll

```diff
@@ -0,0 +1,79 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --include-generated-funcs --version 5
+; RUN: opt -passes="ipsccp<func-spec>" -funcspec-min-function-size=1       \
+; RUN:                                 -funcspec-for-literal-constant=true \
+; RUN:                                 -funcspec-min-codesize-savings=50   \
+; RUN:                                 -funcspec-min-latency-savings=0     \
+; RUN:                                 -S < %s | FileCheck %s
+
+; Verify that we are able to estimate the codesize savings by looking through
+; calls to ssa_copy intrinsics, which are inserted by PredicateInfo when IPSCCP
```

Comment:
I am not seeing any ssa copy instrincis here. Perhaps you want to get the IR after SCCP has added them?

Reply:
The ssa.copy intrinsics are added during IPSCCP when it invokes PredicateInfo:
https://github.com/llvm/llvm-project/blob/22081dc40ba8b13e70df71bb7350126658e8cc70/llvm/lib/Transforms/IPO/SCCP.cpp#L130

I think that this test is a reasonable test case, as its a realistic-ish example of where the intrinsics would be added. Additionally, we cannot insert them ourselves into the test IR - the SCCPSolver asserts() that it has PredicateInfo available for every ssa.copy call in the IR of a function (i.e. it expects that all ssa.copy intrinsics were added by the call to `addPredicateInfo`:
https://github.com/llvm/llvm-project/blob/d924a9ba03a05b417676e84f6c81aac44f907f71/llvm/lib/Transforms/Utils/SCCPSolver.cpp#L1852-L1855

I'm not sure that there's another way to test this, even in a unit-test - unless you have any ideas?

Reply:
Perhaps I'm missing something, but I don't see how it would be possible to write a unit test for this with explicit calls to `@llvm.ssa.copy` intrinsics, for the same reason that we can't do this in the regression tests. 

If we write a unit test such as the following:
```
TEST_F(FunctionSpecializationTest, SSACopy) {
  const char *ModuleString = R"(
    define i32 @foo(i32 %x, i32 %y) {
    entry:
      %x.copy = call i32 @llvm.ssa.copy.i32(i32 %x)
      %add.x.y = add i32 %x.copy, %y
      %cond = icmp eq i32 %add.x.y, 2
      br i1 %cond, label %cond.true, label %cond.false
    cond.true:
      ret i32 %x
    cond.false:
      ret i32 %y
    }
  )";

  Module &M = parseModule(ModuleString);
  Function *F = M.getFunction("foo");
  FunctionSpecializer Specializer = getSpecializerFor(F);
  InstCostVisitor Visitor = Specializer.getInstCostVisitorFor(F);
...
```
We have to construct the InstCostVisitor via getSpecializerFor(F), which constructs an IPSCCPSolver and runs it on the Module (as the Solver is necessary for the InstCostVisitor to function). Therefore we will hit the assert regarding predicate info:
`void llvm::SCCPInstVisitor::handleCallResult(CallBase &): Assertion `PI && "Missing predicate info for ssa.copy"' failed.`

I could add a unit test with code similar to the regression test, that relies on IR that we know will trigger creation of the `@llvm.ssa.copy` intrinsics, but that seems perhaps redundant? Apologies if I'm missing something!

Reply:
Good point. Perhaps then we can keep this regression test and add another RUN line to check predication info?

`; RUN: opt -passes=print-predicateinfo -disable-output -S < %s 2>&1 | FileCheck %s --check-prefix=PREDINFO`

We can add `--check-prefix=FUNCSPEC` to the existing RUN line.

Reply:
Ah nice, good idea. Have updated the regression test to incorporate the suggestions :)